[WIP] Two step admission deocumentation (#1181)
kubernetes-sigs/kueue: site/content/en/docs/concepts/admission_check.md

---
title: "Admission Check"
date: 2023-10-05
weight: 6
description: >

>   A mechanism allowing custom internal or external components to control the timing of

Reviewer: ```suggestion
  A mechanism allowing internal or external components to influence the timing of
```

Author: done

  workloads admission.

> ---

Reviewer: Let's add a paragraph here that explains the high level idea:

AdmissionChecks are a mechanism that allows Kueue to consider additional criteria before admitting a Workload. When a ClusterQueue has AdmissionChecks configured, each of the checks has to provide a
positive signal to the Workload before it can be admitted (link to the admission definition).
You can configure the mechanism using the following components:

Author: done

## Components

### Admission Check

> Admission check is a none namespaced API object used to define details about an `AdmissionCheck` like:

Reviewer: ```suggestion
Admission check is a non-namespaced API object used to define details about an `AdmissionCheck` like:
```

Author: done

#### controllerName

> Is the name with which controller identifies with, not necessarily a K8S Pod or Deployment name. Cannot be empty.

Reviewer: ```suggestion
It's an identifier for the controller that processes this AdmissionCheck, not necessarily a Kubernetes Pod or Deployment name. Cannot be empty.
```

Author: done

#### retryDelayMinutes

Specifies how long to keep the workload suspended after a failed check (after it transitioned to False). After that the check state goes to "Unknown". The default is 15 min.

#### parameters

> Identifies the resource providing additional check parameters.

Reviewer: I think it would be easier to read if we just use a single unnumbered list for the description of all fields.

Author: done

> Identifies the resource providing additional check parameters.

Reviewer: ```suggestion
Identifies an additional resource providing additional parameters for the check.
```

Author: done

> ```yaml

Reviewer: "An AdmissionCheck object looks like the following:"

Author: done

apiVersion: kueue.x-k8s.io/v1beta1
kind: AdmissionCheck
metadata:
  name: prov-test
spec:
  controllerName: kueue.x-k8s.io/provisioning-request
  retryDelayMinutes: 15
  parameters:
    apiGroup: kueue.x-k8s.io
    kind: ProvisioningRequestConfig
    name: prov-test-config
```

### ClusterQueue admissionChecks

> Once defined an AdmissionCheck can be referenced in the ClsterQueues Spec resulting in all Workloads associated with the queue needing to be evaluated by its controller before being admitted. Similarly with `ResourceFalvors`, if an `AdmissionCheck` is not found or its controller has not mark it as `Active`, the cluster queue will be marked as Inactive.

Reviewer: ```suggestion
Once defined, an AdmissionCheck can be referenced in the ClusterQueues' spec. All Workloads associated with the queue need to be evaluated by the AdmissionCheck's controller before being admitted.

Similarly to `ResourceFlavors`, if an `AdmissionCheck` is not found or its controller has not mark it as `Active`, the cluster queue will be marked as Inactive.
```

Author: done

### AdmissionCheckState

> AdmissionCheckState is the way the way the state of an AdmissionCkeck for a specific Workload is tracked.

Reviewer: ```suggestion
AdmissionCheckState is the way the state of an AdmissionCkeck for a specific Workload is tracked.
AdmissionCheckStates are listed in the Workload's `.status.admissionCheckStates` field.
```

Author: done

It is following the schema:
```yaml

> properties:

Reviewer: Let's have an example instead.

"The status of a Workload that has pending AdmissionChecks looks like the following:"

Author: done

Author: done

      using the time when the API field changed is acceptable.
    format: date-time
    type: string
  message:
    description: message is a human readable message indicating
      details about the transition. This may be an empty string.
    maxLength: 32768
    type: string
  name:
    description: name identifies the admission check.
    maxLength: 316
    type: string
  podSetUpdates:
    items:
      description: PodSetUpdate contains a list of pod set modifications
        suggested by AdmissionChecks. The modifications should be
        additive only - modifications of already existing keys or
        having the same key provided by multiple AdmissionChecks
        is not allowed and will result in failure during workload
        admission.
  <...>
      type: object
    type: array
    x-kubernetes-list-type: atomic
  state:
    description: status of the condition, one of True, False, Unknown.
    enum:
    - Pending
    - Ready
    - Retry
    - Rejected
    - PreemptionRequired
    type: string
required:
- lastTransitionTime
- message
- name
- state
type: object
```

A list of states being maintained in the Status of all the monitored Workloads.

> Kueue ensurers that the list of the Workloads AdmissionCheckStates is in sync with the list of its associated ClusterQueue, new checks being added with the `Pending` state.

Reviewer: ```suggestion
Kueue ensures that the list of the Workloads AdmissionCheckStates is in sync with the list of its associated ClusterQueue. Kueue adds new checks with the `Pending` state.
```

Author: done

> - Once a workload has QuotaReservation and all its AdmissionChecks are in "Ready" state it will become Admitted.

Reviewer: Let's add a section in the Workload page for the important conditions (Admitted, QuotaReserved) and add a link to it here.

Author: The workload is described as  " unit of admission" linked to glossary that will link to here. 

I see the conditions more like an implementation detail, and and since no other part of the workload status is described there, I'd say to leave it as is.

Reviewer: I don't think it's an implementation detail. It's the canonical way how Kueue communicates (to other controllers and end users) that a Workload is admitted.

- If at least one of the Workloads AdmissionCheck is in the `Retry` state.
  - If `Admitted` the workload is evicted.
  - If the workload has `QuotaReservation` it will be release released.
- If at least one of the Workloads AdmissionCheck is in the `Rejected`:
  - If `Admitted` the workload is evicted.
  - If the workload has `QuotaReservation` it will be release released.
  - The workload is marked as 'Finished' with a relevant failure message.

> - `PreemptionRequired` is a state specific to PreemptionPolicies, from an Admission point of view being synonym to `Pending`

Reviewer: This was not added yet, right?

Author: yes, removed

### Admission Check Controller

Is a component that monitors Workloads maintaining the content of its specific `AdmissionCheckStates` and the `Active` condition of the `AdmissionCheck`s  its  controlling.

> The logic based on which an `AdmissionCheck` changes states in not in the scope of Kueue.

Reviewer: ```suggestion
The logic for how an `AdmissionCheck` changes states is not part of Kueue.
```

Author: done

### Admission Check

Admission check is a non-namespaced API object used to define details about an `AdmissionCheck` like:

- **controllerName** - It's an identifier for the controller that processes this AdmissionCheck, not necessarily a Kubernetes Pod or Deployment name. Cannot be empty.
- **retryDelayMinutes** - Specifies how long to keep the workload suspended after a failed check (after it transitioned to False). After that the check state goes to "Unknown". The default is 15 min.
- **parameters** - Identifies an additional resource providing additional parameters for the check.

An AdmissionCheck object looks like the following:
```yaml
apiVersion: kueue.x-k8s.io/v1beta1
kind: AdmissionCheck
metadata:
  name: prov-test
spec:
  controllerName: kueue.x-k8s.io/provisioning-request
  retryDelayMinutes: 15
  parameters:
    apiGroup: kueue.x-k8s.io
    kind: ProvisioningRequestConfig
    name: prov-test-config
```

### ClusterQueue admissionChecks

Once defined, an AdmissionCheck can be referenced in the ClusterQueues' spec. All Workloads associated with the queue need to be evaluated by the AdmissionCheck's controller before being admitted.

> Similarly to `ResourceFlavors`, if an `AdmissionCheck` is not found or its controller has not mark it as `Active`, the cluster queue will be marked as Inactive.

Author: done

  <...>
      type: object
    type: array
    x-kubernetes-list-type: atomic
  state:
    description: status of the condition, one of True, False, Unknown.
    enum:
    - Pending
    - Ready
    - Retry
    - Rejected
    - PreemptionRequired
    type: string
required:
- lastTransitionTime
- message
- name
- state
type: object
```

A list of states being maintained in the Status of all the monitored Workloads.

Kueue ensures that the list of the Workloads AdmissionCheckStates is in sync with the list of its associated ClusterQueue, Kueue adds new checks with the `Pending` state.

- Once a workload has QuotaReservation and all its AdmissionChecks are in "Ready" state it will become Admitted.
- If at least one of the Workloads AdmissionCheck is in the `Retry` state.
  - If `Admitted` the workload is evicted.
  - If the workload has `QuotaReservation` it will be release released.
- If at least one of the Workloads AdmissionCheck is in the `Rejected`:
  - If `Admitted` the workload is evicted.
  - If the workload has `QuotaReservation` it will be release released.
  - The workload is marked as 'Finished' with a relevant failure message.

### Admission Check Controller

> Is a component that monitors Workloads maintaining the content of its specific `AdmissionCheckStates` and the `Active` condition of the `AdmissionCheck`s  its  controlling.

Author: done

> The logic for how an `AdmissionCheck` changes states is not part of Kueue.

Author: done
